include/pybind11/stl.h

```diff
@@ -228,6 +229,26 @@ template <typename Key, typename Value, typename Compare, typename Alloc> struct
 template <typename Key, typename Value, typename Hash, typename Equal, typename Alloc> struct type_caster<std::unordered_map<Key, Value, Hash, Equal, Alloc>>
   : map_caster<std::unordered_map<Key, Value, Hash, Equal, Alloc>, Key, Value> { };
 
+/// Helper class to reset an std::optional-like class in the best way.
+template<typename T> struct optional_reset {
+private:
+    // Either or both of these overloads may exist (SFINAE decides), and if
+    // both exist then the .reset version is preferred due to a better match
+    // in the second argument.
+    template<typename T2 = T>
+    static decltype(std::declval<T2>().reset()) reset_impl(T &value, int) {
```

Comment:
`static void_t<decltype(std::declval<T2>().reset())> reset_impl...` would be better here.